### ISSUE TITLE
Add logic to reset flags for a crop column that just became inactive

### ISSUE THREAD

**Issue by billsacks** [ESCOMP/ctsm#209]
_From @billsacks on November 3, 2017 20:37_

Currently, `croplive` is only updated for active crop columns (and this is probably true for other similar flags as well). I can envision how this might cause problems for crop columns that start out active, become inactive (due to their gridcell area going to 0), and then later become active again (due to their gridcell area becoming greater than 0). In this situation, `croplive` will start out as true when the column re-activates. I imagine that all sorts of oddities may result in this first year after re-activation.

There probably aren't many places where this happens in practice, so this may not be a big issue, but it's probably still worth fixing.

Some ways I can think of to fix this:

(1) Have some special logic in subgridWeightsMod: is_active_c (and/or is_active_p) that keeps a crop column/patch active for one time step after the time at which it first becomes inactive (i.e., if its weight was > 0 in the _last_ timestep). Crop columns / patches that are about to become inactive (i.e., would be inactive if it weren't for this special logic) would get a flag set (e.g., `crop_newly_inactive`). Then, in the crop phenology routine, any crop patch with `crop_newly_inactive` True (and `croplive` currently True) would go through the harvest logic, in order to set `croplive` to False, etc.

Notes about this:

- We need to keep the column active for this extra time step so that some other cleanup steps can happen. One example that I'm aware of is that we need the logic to reset GDDPLANT to accumResetValue to be triggered; with some changes I'm planning to make very soon, that would only happen properly if the patch were active.

- I'd like to have this new flag (`crop_newly_inactive`) rather than, e.g., checking (active and weight = 0). This is because I want to keep the capability to maintain crop columns as always-active, in which case they shouldn't go through this instant-harvest logic. Keeping the `crop_newly_inactive` logic in subgridWeightsMod makes it easier to change this to keep crop columns always active, if desired.

(2) Alternatively, we could do this via resetting the appropriate flags when a column/patch newly becomes active. For columns, we could potentially do this in dynInitColumnsMod, although (a) This isn't currently set up to reset flags for a newly-active column - though in principle it could be, and (b) We currently don't have analogous infrastructure for new patches.

In addition to setting `croplive` (and possibly other similar flags), we may want to have this reset anything – like the GDDPLANT accumulator – that would normally get reset when croplive is .false., etc. On the other hand, resetting all this extra stuff could be a maintenance problem, and it may be okay to just rely on the existing code to do this. (e.g., for GDDPLANT, it will get reset as long as croplive stays .false. for at least one time step... it may not be completely fool-proof to rely on this, but it may be good enough.)

_Copied from original issue: NCAR/CLM#12_

**Comment by billsacks** [comment]
Another way to handle this could be:

(3) Gather all code that gets executed when `croplive` becomes `.false.` into a single routine - or at least have it coordinated by a single routine (which may make calls out to the appropriate places). Then have the dynamic landunits code call this routine if it detects a crop column that has newly become inactive.

I like this solution because it avoids the complexity of keeping a column active for an additional time step (as is needed in (1)). I also like the idea of making it clear what happens when a crop newly becomes non-live.

**Comment by samsrabin** [comment]
I have also run into this and was planning to make my own issue, but saw this. I'll just bump it here: _bump_

**Comment by billsacks** [comment]
@samsrabin - thanks for your comment. Are there particular situations in which you have encountered this and/or problems it has caused that could be helpful to add more context / motivation for this issue?

**Comment by samsrabin** [comment]
Yes indeed! Sorry I didn't do so before, but I didn't want to take any real time out of the meeting this morning.

Comparing model output yields to observations (such as the FAOSTAT data) is tricky for a number of reasons. One of these is how "year" is reported. There's an FAOSTAT standard but it's unclear how well it's obeyed. The standard for analyses in the Global Gridded Crop Model Intercomparison is to treat the time axis of our outputs as "growing seasons" instead of calendar years. For example, some gridcell's corn growing season 2007 is the corn growing season that started in 2007, regardless of when harvest happened.

This is how I'm doing the analyses for the crop calendars manuscript (#1863; generally trying to stick to GGCMI analysis methods as closely as possible). I thus added a "year of planting" output to CLM that allows calendar-annual crop outputs (with my new "number of harvests" dimension) to be assigned a growing season number in postprocessing. 

The code I wrote for that postprocessing was giving me fits until I traced it to the problem you outlined above—old crops become active again. I've been calling these "zombie crops," and my postprocessing code now detects and tosses them during import.

It looks like this happens in 246 patch-years in 1961–2013 with the CLM default crop calendars, so it's not a huge deal, but it is a bug.

Your solution (3) is how I was thinking of handling the issue—something like a `HarvestCrop()` subroutine to be the counterpart of `PlantCrop()`.

I'm wondering how C conservation isn't broken by zombie crops, actually; I'd need to dig in more to what's happening in the land-use transition code to understand what's going on. It could be that zombie crops don't actually end up with any biomass once resurrected.

**Comment by billsacks** [comment]
Thanks for your very clear description. I like the imagery of zombie crops :-)

**Comment by billsacks** [comment]
Oh, and as for carbon conservation: I think this is the key line of code that takes care of this:

https://github.com/ESCOMP/CTSM/blob/05fb17bff05aaf0544170ecbb41984bdbb875807/src/dyn_subgrid/dynPatchStateUpdaterMod.F90#L285

This subroutine is called for each C & N state variable when there are area changes. This line rescales the state variables when there is a growth in area. If the original area was 0 and the new area is greater than 0, then the new state variable will be 0, regardless of its original value.
